app/knight.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import annotations
+
+from app.ammunition.armour import Armour
+from app.ammunition.potion import Potion
+from app.ammunition.weapon import Weapon
+
+
+class Knight:
+    __all_knights = []
+
+    def __init__(
+            self,
+            name: str,
+            power: int,
+            hp: int,
+            armour: [Armour],
+            weapon: Weapon,
+            potion: Potion,
+    ) -> None:
+
+        self.__name = name
+        self.__hp = hp
+        self.__power = power
+        self.__protection = 0
+        self.__weapon = weapon
+        self.__armour = armour
+        self.__potion = potion
+
+        Knight.__all_knights.append(self)
+
+    def get_name(self) -> str:
+        return self.__name
+
+    def get_power(self) -> int:
+        return self.__power
+
+    def get_hp(self) -> int:
+        return self.__hp
+
+    def get_protection(self) -> int:
+        return self.__protection
+
+    def get_armour(self) -> [Armour]:
+        return self.__armour
+
+    def get_weapon(self) -> Weapon:
+        return self.__weapon
+
+    def get_potion(self) -> [Potion]:
+        return self.__potion
+
+    def __apply_armour(self) -> None:
+        self.__protection += sum([armour.get_protection()
+                                  for armour in self.__armour])
+
+    def __apply_potion(self) -> None:
+        effects = self.__potion.get_effect()
+        # print(effects)
+        if "power" in effects:
+            self.__power += effects["power"]
+        if "hp" in effects:
+            self.__hp += effects["hp"]
+        if "protection" in effects:
+            self.__protection += effects["protection"]
+
+    def __appy_weapon(self) -> None:
+        self.__power += self.__weapon.get_power()
+
+    def prepare_for_battle(self) -> None:
+        self.__apply_armour()
+        self.__appy_weapon()
+        try:
+            self.__apply_potion()
+        except AttributeError:
+            pass
+
+    def hit(self, other: Knight) -> None:
+        if self.__power >= other.__protection + other.__hp:
+            other.__protection = 0
+            other.__hp = 0
+
+        elif self.__power >= other.__protection:
+            other.__hp -= self.__power - other.__protection
+            other.__protection = 0
```

Comment:
I really like your classes, so why dont you use all those getters here in this method? :)

Reply:
i'm not sure that we need to use getters out of a class, cause we have access in a class to variables directly. I've not readed about it yet, so if you have some information/article which refutes my approach of code. I'll read it with great plesaure

Reply:
in fact i just wanted to point out that it seemed strange that after you've implemented in Knight methods like "get_armour", "get_hp", "get_power" etc they havent been used here. I mean it is ok to address attributes on self but it would be cool to get attribute of other via those methods. Anyway that was more like suggestion so dont mind if you dont feel like it.

Reply:
It seems to me it isn't a bad idea leave getters there. Now they are unnecessary. But suppose that in a real project someone want in the future get name/power etc of a knight. Without getters he'll need to realize them. such methods like get_name/get_hp/get_power will use often I think. It's like an example with battle method. You've wrote about potential necessity to change opponents.  According to task description we don't need to change them. But in future can, here the same

Also noticed that IDLE doesn't swear that I don't use them

Reply:
oh no, i havent asked you to remove getters :) i like them too , i just suggested that you could have used them to get the necessary info from the inside of instance passed into hit() method. For example:    if self.__power >= other.get_power() + other.get_hp()  -- like that.

Reply:
I just don't understand why do we need to use  ```self``` like ```self.__power``` and ```other``` like ```other.get_power()```
We use getters outside cause it isn't recommended touch protected/private attributes/methods directly
But in this example ```self``` and ```other``` inside of their class and python allows us to use them directly
So why do we need  touch the attribute of ```self``` directly, and an attribute of ```other``` through getter? Maybe do you have any example where it's shown, for familiarization and analyzing

Reply:
im sorry i dont have any example. I get your point so i wont insist on changing this part here. I guess its just something i would do but that doesnt mean it is the only correct way.